Update repositories.md

jonapich opened the pull request:
# Pull Request Check List

Resolves: some discussion in #3855 

<!-- This is just a reminder about the most common mistakes. Please make sure that you tick all *appropriate* boxes.  But please read our [contribution guide](https://python-poetry.org/docs/contributing/) at least once, it will save you unnecessary review cycles! -->

- [ ] Added **tests** for changed code.
- [ ] Updated **documentation** for changed code.

<!-- If you have *any* questions to *any* of the points above, just **submit and ask**!  This checklist is here to *help* you, not to deter you from contributing! -->

abn commented on docs/repositories.md:
```diff
@@ -112,9 +112,27 @@ default = false
 {{% warning %}}
 
-If package sources are defined for a project, these will take precedence over
+Reminder: If package sources are defined for a project, these will take precedence over
```
I think this is redundant over the existing warning box.

abn commented on docs/repositories.md:
````diff
@@ -112,9 +112,27 @@ default = false
 secondary = false
 ```
 
+When a package source is defined like this, it takes precedence over [PyPI](https://pypi.org).
````
This is redundant I feel as this is said just below as well.

abn commented on docs/repositories.md:
````diff
@@ -112,9 +112,27 @@ default = false
 
+When a package source is defined like this, it takes precedence over [PyPI](https://pypi.org).
+
+To ignore [PyPI](https://pypi.org) completely, use `default = true`.
+
+To target a package source only on specific packages, use `secondary = true` on the package source, and specify the `source` when adding the dependency.
+In the example below, `my-lib` will be obtained from the custom package source and `urllib3` will use the [PyPI](https://pypi.org) source.
+
+```toml
+[tool.poetry.requirements]
+my-lib = { version = "*", source = "foo" }
+urllib3 = "*"
+
+[[tool.poetry.source]]
+name = "foo"
+url = "https://foo.bar/simple/"
+secondary = true
+```
````
I'd move this below the warning for continuity. And refer to the default section. Rephrase it something like -

```md
{{% note %}}
If you prefer to disable PyPI, you may choose to set one of your package sources to be the [default](#default-package-source).
{{% /note %}}
```

And move the exampe to the secondary package source section maybe?

jonapich commented:
@abn I changed a few things. I added the `--secondary` to the first example too since that's what the wording was implying. 

However, this part in the secondary source section is basically false:

```
All package sources (including secondary sources) will be searched during the package lookup process. These network requests will occur for all sources, regardless of if the package is found at one or more sources.

If you wish to avoid this, you may explicitly specify which source to search in for a particular package.
```

You do not "avoid this" by adding the source. If you add the source to a package, the other dependencies _still use the secondary sources_. I would rephrase to:

```
You can force a dependency to use a specific package source (exclusively? or it will look it up first? not sure) by specifying which source to search in for a particular package.
```

jonapich commented on docs/repositories.md:
```diff
@@ -112,23 +119,26 @@ default = false
-[certificates](#certificates), please refer to the relevant sections below.
+If you prefer to disable [PyPI](https://pypi.org) completely, you may choose to set one of your package sources to be the [default](#default-package-source).
+
+To enable a package source only for a specific dependency, see [Secondary Package Sources](#secondary-package-sources).
```
I think this suggestion makes it unclear what this is about. I rephrased to:

> If you prefer to specify a package source for a specific dependency, see [Secondary Package Sources](#secondary-package-sources).

neersighted commented on docs/repositories.md:
```diff
@@ -171,7 +181,18 @@ If you wish to avoid this, you may explicitly specify which source to search in
 package.
```
`source =` should restrict this to only that repository -- if we're doing otherwise, that's a bug.

jonapich replied:
It's difficult for me to assert this, because my private `pypi` is set to redirect to `pypi.org` on missing packages.

Since this closely touches security, there really should be a unit test that asserts it.

neersighted commented on docs/repositories.md:
```diff
@@ -112,23 +119,26 @@ default = false
-sources to be [secondary](#secondary-package-sources).
-
-{{% /warning %}}
+Any package source defined like this takes precedence over [PyPI](https://pypi.org).
```
```suggestion
Any package source not marked as `secondary` will take precedence over [PyPI](https://pypi.org).
```

neersighted commented on docs/repositories.md:
```diff
@@ -167,11 +177,21 @@ All package sources (including secondary sources) will be searched during the pa
 
-If you wish to avoid this, you may explicitly specify which source to search in for a particular
-package.
+In order to limit the search for a specific package to a particular package source, you can explicitly specify what source to use.
```
```suggestion
In order to limit the search for a specific package to a particular package repository, you can specify the source explicitly. This is strongly suggested for all private packages to avoid dependency confusion attacks.
```
